### PR TITLE
bench(transport/recovery): benchmark `SentPackets::take_ranges`

### PR DESCRIPTION
**Pull request by mxinden**
This pull request extracts the benchmark from https://github.com/mozilla/neqo/pull/2245.

I suggest we merge here first. We can then compare https://github.com/mozilla/neqo/pull/2245 against its baseline on `main`.

Note the second separate commit, addressing clippy lints only (82ef97e24c5af3d5098f5b73bcd23c72a7b365b3).

### CODE REVIEW

File: neqo-transport/src/recovery/mod.rs [hunk 1]
```diff
@@ -505,9 +522,14 @@ impl LossRecovery {
         self.qlog = qlog;
     }
 
+    /// Drop all 0rtt packets.
+    ///
+    /// # Panics
+    ///
+    /// Panics when the largest acknowledged or `loss_time` is already set. In
+    /// addition, the client should not have received any ACK frames when it
+    /// drops 0-RTT.
```
Comment (martinthomson) [inline neqo-transport/src/recovery/mod.rs:531]:
```suggestion
    /// Panics when the largest acknowledged or `loss_time` is already set.
    /// The client should not have received any ACK frames in the
    /// application data packet number space when it drops 0-RTT.
```

File: neqo-transport/src/recovery/sent.rs [hunk 2]
```diff
@@ -61,41 +62,54 @@ impl SentPacket {
...
     /// The length of the packet that was sent.
+    #[must_use]
     pub const fn len(&self) -> usize {
         self.len
     }
 
+    #[must_use]
+    pub const fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
```
Comment (martinthomson) [inline neqo-transport/src/recovery/sent.rs:109]: Why did you add this?  Packets cannot be empty, so it doesn't make sense to have this.  You should suppress the warning instead.

File: neqo-transport/src/recovery/sent.rs [hunk 3]
```diff
@@ -174,6 +194,8 @@ pub struct SentPackets {
 }
 
 impl SentPackets {
+    #[allow(clippy::len_without_is_empty)]
```
Comment (martinthomson) [inline neqo-transport/src/recovery/sent.rs:197]: Is this the type you wanted to add an `is_empty()` function for?

**Review by martinthomson — APPROVED** [review]
It's possible that I'm confused.  I'm sure that you'll work it out.

File: neqo-transport/src/recovery/sent.rs (hunk 2 above)
Comment (mxinden) [reply neqo-transport/src/recovery/sent.rs]: Yes. Sorry about that. Fixed with https://github.com/mozilla/neqo/pull/2253/commits/1ed1ca09e6dccb7685c3b88db2afde943247d379.

File: neqo-transport/src/recovery/sent.rs (hunk 3 above)
Comment (mxinden) [reply neqo-transport/src/recovery/sent.rs:193]: The lint only complains, because `SentPackets` is now `pub` when `bench` feature is enabled.

Given that we don't need `SentPackets::is_empty` anywhere within the `neqo_transport::recovery` `mod` (i.e. parent of `sent` `mod`), I think suppressing is the better option here.
